File: JavaSource/src/main/java/ru/akhcheck/patterns/composite/Dish.java
package ru.akhcheck.patterns.composite;

import java.util.ArrayList;

public class Dish implements DishComponent{
    private ArrayList<DishComponent> components = new ArrayList<>();

    public int GetCalories() {
        int calories = 0;
        for (DishComponent component: components) {
            calories += component.GetCalories();
        }
        return calories;
    }

    public Dish add(DishComponent component) {
        components.add(component);
        return this;
    }

}

File: Composite/python-source/ration/Dish.py
from __future__ import annotations

from .DishComponent import DishComponent


class Dish(DishComponent):
    """Узел дерева: держит части и складывает их калории.

    Важно, что части хранятся как DishComponent, а не как Dish или Milk. Узел
    не знает, лист перед ним или ещё один узел, и складывает одинаково.
    """

    def __init__(self) -> None:
        self._components: list[DishComponent] = []

    def add(self, component: DishComponent) -> Dish:
        """Возвращает себя, чтобы вызовы складывались в цепочку."""
        self._components.append(component)
        return self

    def get_calories(self) -> int:
        return sum(component.get_calories() for component in self._components)
